massgov/mayflower-patternlab
# DP-6517 create ajax pattern to allow client side rendering of other patterns

## Description

This change provides a jQuery plugin to process a given pattern data. It provides the option of using a transform method between an endpoint and the TwigJS render process so that if the endpoint cannot be updated to provide exactly what the component needs, it can be transformed.

Based on prior work in #638 + #643 

## Related Issue / Ticket

- [DP-6517](https://jira.state.ma.us/browse/DP-6517)

## Steps to Test


1. [ ] The new organism: ajax pattern renders emergency alerts content + has documentation
    1. Browse to: https://jesconstantine.github.io/mayflower/?p=organisms-ajax-pattern
    1. Verify that it appears and renders like the emergency alerts organism: https://jesconstantine.github.io/mayflower/?p=organisms-emergency-alerts 
    1. Click the cog in the top-right corner, then click show pattern info and verify that the pattern is documented
    1. Devs: verify that the documentation makes sense (i.e. that you might be able to implement it)
    1. Devs: open your browser dev tools to sniff traffic and verify that there is a get request to:
        1. Pattern json data: https://jesconstantine.github.io/mayflower/assets/data/emergency-alerts.json
        1. The pattern twig templates: https://jesconstantine.github.io/mayflower/assets/patterns/03-organisms/by-template/emergency-alerts.twig etc.
1. [ ] Multiple instances of the pattern can be used on one page, rendering different patterns
    1. Browse to: https://jesconstantine.github.io/mayflower/?p=pages-location-park-content
    1. Verify that the emergency (site) alerts render in the header
    1. Verify that the page alerts render just below the nav menu
1. [ ] The rendered pattern js functionality continues to work
    1. [ ] Accordion button works
        1. Browse to: https://jesconstantine.github.io/mayflower/?p=organisms-ajax-pattern    
        1. Verify the hide / show button works (it collapses and expands to show alert messages)
    1. [ ] Accordion state is remembered across pages
        1. Make sure the emergency alerts accordion is in a closed state
        1. Browse to: https://jesconstantine.github.io/mayflower/?p=pages-location-park-content
        1. Verify that the state of the emergency alerts accordion is in a closed state
        1. On the location page, expand the alerts accordion
        1. Refresh or browse back to the ajax pattern organism: https://jesconstantine.github.io/mayflower/?p=organisms-ajax-pattern
        1. Verify that the state of the emergency alerts accordion is expanded
1. [ ] Client side rendered patterns can have their data transformed (and / or overridden) when ajax pattern
    1. Browse to https://jesconstantine.github.io/mayflower/?p=pages-guide
    1. Verify that alerts render as information alerts and the group title is "This alert title was overridden by a custom transform function."
    1. Devs: open your browser dev tools to sniff traffic and verify that the get request is made to the same data json file as the ajax pattern organism but that the content has been overridden
        1. Verify get request for pattern json data: https://jesconstantine.github.io/mayflower/assets/data/emergency-alerts.json
        1. Browse to https://jesconstantine.github.io/mayflower/assets/data/emergency-alerts.json
        1. Confirm that `emergencyAlerts.emergencyHeader.title` is "The State is experiencing severe weather due to the winter storm Paula." and not "This alert title was overridden by a custom transform function."
        1. Confirm that `emergencyAlerts.emergencyHeader.prefix` is not set (default is Emergency Alerts)
1. [ ] Alerts can render as emergency alerts and / or notices (informational alerts)
    1. Site alerts
        1. Browse to https://jesconstantine.github.io/mayflower/?p=organisms-emergency-alerts-as-notice
        1. Observe the "Informational Alert" prefix in the dark gold banner
        1. Click the cog in the top-right corner, then click show pattern info and verify that the pattern variant is documented and it is clear how to create the variant
    1. Page alerts
        1. Browse to https://jesconstantine.github.io/mayflower/?p=organisms-header-alert-as-notice
        1. Observe the "Notice" prefix in the dark gold banner
        1. Click the cog in the top-right corner, then click show pattern info and verify that the pattern variant is documented and it is clear how to create the variant
1. [ ] Alerts can be rendered as a full page
    1. Site alerts: https://jesconstantine.github.io/mayflower/?p=pages-site-alert
    1. Page alerts: https://jesconstantine.github.io/mayflower/?p=pages-page-alert
    1. Template content type: https://jesconstantine.github.io/mayflower/?p=templates-alert

   
## Screenshots
Use something like [licecap](http://www.cockos.com/licecap/) to capture gifs to demonstrate behaviors.


## Additional Notes:

Anything else to add?

#### Impacted Areas in Application


* Any pattern which has an existing js module and needs to be rendered clientside (because the js needs to be touched to both initialize a pattern on DOM load and also listing for a custom `MA:AjaxPattern:Render` event and then initialize the newly rendered pattern.

#### @TODO


* Validate data (after transform if that function is passed) against a json schema
* Implement an aria live solution (like Drupal.announce) for dynamic content

#### Today I learned...


- Some libraries (like TwigJS) require function names to be retained during uglify/minification and can therefor have their functionality broken when using tools like gulp-uglify and not passing config to not mangle function names. See [uglify mangle options](https://github.com/mishoo/UglifyJS2#mangle-options).
